jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala

```diff
@@ -431,126 +213,97 @@ object XGBoost extends XGBoostStageLevel {
+      // So we need to throw the real exception instead of shutdown error.
+      try {
+        Communicator.shutdown()
+      } catch {
+        case e: Throwable =>
+          logger.error("Communicator.shutdown error: ", e)
+          throw previousException.getOrElse(e)
+      }
     }
   }
 
   /**
-   * @return A tuple of the booster and the metrics used to build training summary
+   * Train a XGBoost booster with parameters on the dataset
+   *
+   * @param input         the input dataset for training
+   * @param runtimeParams the runtime parameters for jvm
+   * @param xgboostParams the xgboost parameters to pass to xgboost library
+   * @return the booster and the metrics
    */
-  @throws(classOf[XGBoostError])
-  private[spark] def trainDistributed(
-      sc: SparkContext,
-      buildTrainingData: XGBoostExecutionParams => (RDD[() => Watches], Option[RDD[_]]),
-      params: Map[String, Any]):
-    (Booster, Map[String, Array[Float]]) = {
+  def train(input: RDD[Watches],
+            runtimeParams: RuntimeParams,
+            xgboostParams: Map[String, Any]): (Booster, Map[String, Array[Float]]) = {
 
-    logger.info(s"Running XGBoost ${spark.VERSION} with parameters:\n${params.mkString("\n")}")
+    val sc = input.sparkContext
+    logger.info(s"Running XGBoost ${spark.VERSION} with parameters: $xgboostParams")
 
-    val xgbParamsFactory = new XGBoostExecutionParamsFactory(params, sc)
-    val runtimeParams = xgbParamsFactory.buildXGBRuntimeParams
+    // TODO Rabit tracker exception handling.
+    val trackerConf = runtimeParams.trackerConf
 
-    val prevBooster = runtimeParams.checkpointParam.map { checkpointParam =>
-      val checkpointManager = new ExternalCheckpointManager(
-        checkpointParam.checkpointPath,
-        FileSystem.get(sc.hadoopConfiguration))
-      checkpointManager.cleanUpHigherVersions(runtimeParams.numRounds)
-      checkpointManager.loadCheckpointAsScalaBooster()
-    }.orNull
-
-    // Get the training data RDD and the cachedRDD
-    val (trainingRDD, optionalCachedRDD) = buildTrainingData(runtimeParams)
+    val tracker = new RabitTracker(runtimeParams.numWorkers,
+      trackerConf.hostIp, trackerConf.port, trackerConf.timeout)
+    require(tracker.start(), "FAULT: Failed to start tracker")
 
     try {
-      val (booster, metrics) = withTracker(
-        runtimeParams.numWorkers,
-        runtimeParams.trackerConf
-      ) { tracker =>
-        val rabitEnv = tracker.getWorkerArgs()
-
-        val boostersAndMetrics = trainingRDD.barrier().mapPartitions { iter =>
-          var optionWatches: Option[() => Watches] = None
-
-          // take the first Watches to train
-          if (iter.hasNext) {
-            optionWatches = Some(iter.next())
+      val rabitEnv = tracker.getWorkerArgs()
+
+      val boostersAndMetrics = input.barrier().mapPartitions { iter =>
+        require(iter.hasNext, "Couldn't get DMatrix")
+        val watches = iter.next()
```

Comment:
It looks like QuantileDMatrix is computed here, but the Communicator is not initialized till inside the trainBooster call below.   So still not fully clear if communication is happening for the QuantileDMatrix per response to my previous comment on this.

Reply:
You're right, it's my bad. the communicator should be initialized before creating quantile dmatrix. I just made sure the local tests on only 1 GPU can pass. This issue definitely needs to be found in multi-GPUs environment.

Reply:
The newest commit has fixed this issue. Thx.